### ISSUE TITLE
Rename Symbol method

### ISSUE THREAD

**Issue by hypersurge** [vshaxe/vshaxe#188]
Rename symbol currently supports local variable and parameter.  When attempting to rename a method (via Rename Symbol menu) "Error" message is displayed (top).

Adding rename method support would be valuable (even if restricted to private methods initially).

**Comment by Gama11** [comment]
Restricting it to private variables wouldn't help, since "private" really means "protected" in Haxe (so you still have to take multiple files into account). Besides, you can use `@:access` metadata to access private fields even outsidse of a subclass.

**Comment by hypersurge** [comment]
Good point.  I am unsure whether a rename method function (limited to a single file) would be of sufficient value to warrant implementing.  So that leaves the harder challenge of identifying all impacted files - I am wondering what other functionality has overlaps?

**Comment by Kyliathy** [comment]
Just curious, why is this "blocked"? Are there serious impediments? I feel that refactoring capabilities are quite important :). Would be nice to have stuff such as extracting a property from a variable :). And to handle renaming of files & imports once a class is renamed.

@Gama11 , how would you estimate the amount of effort necessary to implement this in vshaxe? I might be up for it.

**Comment by Gama11** [comment]
It's blocked because we can't really implement this in vshaxe / haxe-language-server - it has to be done through a new method in Haxe's display API. Only the compiler has the information necessary to implement this correctly.

**Comment by Kyliathy** [comment]
Is there anybody that can assist with this? Renaming is one of the most common and frequent activities. I feel this will improve adoption for Haxe. Of course, I can use IntelliJ to refactor... is that how you are doing it?

**Comment by Gama11** [comment]
Well, @simn is the one who knows the OCaml / compiler side of things best.

No, I just tend to use find-and-replace. I never found FlashDevelop's rename feature very reliable for Haxe, I imagine IntelliJ is similar. I agree it's a very important feature though.

**Comment by joshtynjala** [comment]
I had an idea for how the language server could actually do this without compiler changes, so I gave it a try, but no success. My hypothesis was that the "find references" feature should be able to return all locations to be renamed. However, when I gave it a try, I discovered that find references fails to return all references in many cases. Only a subset. Oh well, worth a try!

For reference, here's the code that I tried (I had to also expose context.findReferences as public):

```hx
context.gotoDefinition.onGotoDefinition(params, token, function(locations) {
	function noDeclaration() {
		reject(ResponseError.internalError("No declaration found."));
	}
	if (locations == null) {
		return noDeclaration();
	}
	final declaration = locations[0];
	if (declaration == null) {
		return noDeclaration();
	}
	final changes = new haxe.DynamicAccess();
	function addEdit(reference:Location) {
		var uriAsString = reference.uri.toString();
		var edits:Null<Array<TextEdit>> = Reflect.field(changes, uriAsString);
		if (edits == null) {
			edits = [];
			changes[uriAsString] = edits;
		}
		edits.push({
			range: reference.range,
			newText: params.newName
		});
	}
	addEdit(declaration);
	context.findReferences.onFindReferences(params, token, function(references) {
		function noDeclaration() {
			reject(ResponseError.internalError("No declaration found."));
		}
		if (references == null) {
			return noDeclaration();
		}
		for (reference in references) {
			addEdit(reference);
		}
		resolve({changes: changes});
	}, _ -> invalidRename());
}, _ -> invalidRename());
```

Really hoping the compiler folks consider prioritizing symbol renaming sooner rather than later. I have seen potential users say that they've chosen not to adopt Haxe because this core feature is missing from VSCode.

**Comment by AlexHaxe** [comment]
early wip: https://github.com/HaxeCheckstyle/haxe-rename/discussions/1

**Comment by AlexHaxe** [comment]
fixed by vshaxe/haxe-languageserver#57
